clients/src/main/java/org/apache/kafka/common/replica/ReplicaSelector.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.common.replica;
+
+import org.apache.kafka.common.Configurable;
+import org.apache.kafka.common.Node;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.security.auth.KafkaPrincipal;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.net.InetAddress;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+public interface ReplicaSelector extends Configurable, Closeable {
+
+    /**
+     * Select the preferred replica a client should use for fetching. If no replica is available, this will return an
+     * empty optional.
+     */
+    Optional<ReplicaInfo> select(TopicPartition topicPartition,
+                                 ClientMetadata clientMetadata,
+                                 Set<ReplicaInfo> replicaInfos);
```

Comment:
Would it make sense to wrap this in a `PartitionInfo` object? That will be easier to evolve if we find the need for new information which is at the partition level instead of the replica level. For example, we could have a separate accessor for the leader and take `isLeader` out of `ReplicaInfo`.

Reply:
Do you mean the return type or the 3rd argument? The problem with taking PartitionInfo as an argument to `select` is it doesn't have some of the info we want like last caught up time and LEO.

As a return type, that would work fine I think since we mainly care about the `Node` for the replica.

Reply:
I was thinking more conceptually, not necessarily reusing `PartitionInfo`. We should have an object which represents the partition state.

Reply:
Ah, I see what you mean, yea that seems like a good idea.